controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/deployment/ApplicationVersion.java

```diff
@@ -39,24 +40,22 @@
     public ApplicationVersion(Optional<SourceRevision> source, OptionalLong buildNumber, Optional<String> authorEmail,
                                Optional<Version> compileVersion, Optional<Instant> buildTime, Optional<String> sourceUrl,
                                Optional<String> commit) {
-        Objects.requireNonNull(source, "source cannot be null");
-        Objects.requireNonNull(buildNumber, "buildNumber cannot be null");
-        Objects.requireNonNull(authorEmail, "author cannot be null");
-        if (source.isPresent() != buildNumber.isPresent()) {
-            throw new IllegalArgumentException("both buildNumber and source must be set together");
-        }
-        if (compileVersion.isPresent() != buildTime.isPresent()) {
-            throw new IllegalArgumentException("both compileVersion and buildTime must be set together");
-        }
-        if (buildNumber.isPresent() && buildNumber.getAsLong() <= 0) {
-            throw new IllegalArgumentException("buildNumber must be > 0");
-        }
-        if (authorEmail.isPresent() && ! authorEmail.get().matches("[^@]+@[^@]+")) {
+        if (buildNumber.isEmpty() && (   source.isPresent() || authorEmail.isPresent() || compileVersion.isPresent()
+                                      || buildTime.isPresent() || sourceUrl.isPresent() || commit.isPresent()))
+            throw new IllegalArgumentException("Build number must be present if any other attribute is");
+
+        if (buildNumber.isPresent() && buildNumber.getAsLong() <= 0)
+            throw new IllegalArgumentException("Build number must be > 0");
+
+        if (commit.isPresent() && commit.get().length() > 128)
```

Comment:
Validate URL somewhere?

Reply:
Hmm... Yes, for the explicitly passed `sourceUrl`, I guess that makes sense. I don't dare do it for the generated ones :)

Reply:
`URI.create()` is a start, but unfortunately it allows relative URLs or just the path component